### PR TITLE
Fix App deprecation check mechanism

### LINKED ISSUE #1235: App deprecation mechanism seems broken
### Expected Behavior

The phone shows the deprecation screen on startup when its version is out of date.
Also, bonus points for making the version checking smarter. Atm we need to register every single version and mark it deprecated in the firebase DB. it should instead just have a single minimum version

### Current Behavior

Volpe's phone didn't show the deprecation notice.

### PR DESCRIPTION
**Pull request by rohit-dua**
### Description

The app deprecation check mechanism was not working as it required firebase login with web3 account. This requirement is not required for the version check.

Atm we need to register every single version and mark it deprecated in the firebase DB. Now a minimum version in firebase db should do the task

### Testing

The deprecation system was tested on the device by:
* Changing the deprecation flag of the testing version (1.5.0) in the firebase db. The format `versions > 1 > 5 > 0 > deprecated > true` was used.
* Adding a `minVersion` flag to versions and providing it a higher value than the current version

In both the cases the deprecation screen was showed at app start.

### Related issues

- Fixes #1235

### Backwards compatibility

It is backwards compatible.

### CODE REVIEW

File: packages/mobile/src/firebase/firebase.ts [hunk 1]
```diff
@@ -114,13 +114,49 @@ export const writePaymentRequest = (paymentInfo: PaymentRequest) => async () =>
   }
 }
 
+function getHigerVersion(version1: string, version2: string): string | undefined {
```
Comment (jmrossy) [inline packages/mobile/src/firebase/firebase.ts]: Typo Higer

File: packages/mobile/src/firebase/firebase.ts [hunk 2]
```diff
@@ -114,13 +114,49 @@ export const writePaymentRequest = (paymentInfo: PaymentRequest) => async () =>
...
+  for (let i = 0; i < minVersionLength; i++) {
+    if (version1Array[i] > version2Array[i]) {
+      return version1
+    } else if (version1Array[i] < version2Array[i]) {
+      return version2
+    }
+  }
+  if (version1Array.length > version2Array.length) {
+    return version1
+  } else if (version1Array.length < version2Array.length) {
```
Comment (jmrossy) [inline packages/mobile/src/firebase/firebase.ts]: And if both arrays are equal?

File: packages/mobile/src/firebase/firebase.ts [hunk 3]
```diff
@@ -114,13 +114,49 @@ export const writePaymentRequest = (paymentInfo: PaymentRequest) => async () =>
...
+  }
+  Logger.info(TAG, `Checking version info ${version}`)
+  const minVersion = (await firebase
+    .database()
+    .ref('versions/minVersion')
+    .once('value')).val()
+  if (!minVersion) {
+    return { version, deprecated }
+  }
+  const minVersionVal = minVersion.val
```
Comment (jmrossy) [inline packages/mobile/src/firebase/firebase.ts]: why the .val here if you have a .val() above on 150?

File: packages/mobile/src/firebase/firebase.ts [hunk 4]
```diff
@@ -114,13 +114,49 @@ export const writePaymentRequest = (paymentInfo: PaymentRequest) => async () =>
...
+  const higherVersion = getHigerVersion(minVersionVal, version)
```
Comment (jmrossy) [inline packages/mobile/src/firebase/firebase.ts]: This seems like an unintuitive way to answer this question. Could you now just make an isVersionDeprecated function instead which returns a boolean?

**Review by jmrossy — COMMENTED** [review]
Please include info about how you tested this

File: packages/mobile/src/firebase/firebase.ts [hunk 5]
```diff
@@ -114,13 +114,45 @@ export const writePaymentRequest = (paymentInfo: PaymentRequest) => async () =>
   }
 }
 
+function isDeprecatedVersion(version: string, minVersion: string): boolean {
```
Comment (jmrossy) [inline packages/mobile/src/firebase/firebase.ts]: Please add some unit tests for this

File: packages/mobile/src/firebase/firebase.ts [hunk 6]
```diff
@@ -114,13 +114,45 @@ export const writePaymentRequest = (paymentInfo: PaymentRequest) => async () =>
...
+  // backward compatible
+  let deprecated = false
   const versionFSPath = version.split('.').join('/')
   Logger.info(TAG, `Checking version info ${version}`)
-  return (await firebase
+  const versionInfo = (await firebase
     .database()
     .ref(`versions/${versionFSPath}`)
     .once('value')).val()
+  if (versionInfo && versionInfo.deprecated !== undefined) {
```
Comment (jmrossy) [inline packages/mobile/src/firebase/firebase.ts]: Can you add a comment above this function about explaining the expected format for the version in in the db

File: packages/mobile/src/firebase/firebase.ts [hunk 7]
```diff
@@ -114,13 +114,45 @@ export const writePaymentRequest = (paymentInfo: PaymentRequest) => async () =>
...
+    return versionInfo
+  }
+  const versionsInfo = (await firebase
```
Comment (jmrossy) [inline packages/mobile/src/firebase/firebase.ts:198]: why do we query twice?

**Review by jmrossy — COMMENTED** [review]
As I asked in my last review, please add info about how this was tested.

**Comment by rohit-dua** [comment]
> As I asked in my last review, please add info about how this was tested.

Added the Testing information in the description:
### Testing

The deprecation system was tested on the device by:
* Changing the deprecation flag of the testing version (1.5.0) in the firebase db. The format `versions > 1 > 5 > 0 > deprecated > true` was used.
* Adding a `minVersion` flag to versions and providing it a higher value than the current version

In both the cases the deprecation screen was showed at app start.

File: packages/mobile/src/firebase/firebase.ts [hunk 8]
```diff
@@ -131,26 +131,44 @@ function isDeprecatedVersion(version: string, minVersion: string): boolean {
...
+  }
+  return nestedDict
+}
+
+/*
+Get the Version deprecation information.
+@param version: string The version to check for deprecation
+@return: object { version: <VERSION>, deprecated: <BOOLEAN> }
+Firebase DB Format: 
+  (Backward Compatibility) Add child to versions category in the format versions>1>5>0>{deprecated: true}
```
Comment (jmrossy) [inline packages/mobile/src/firebase/firebase.ts]: As discussed, we won't need this. Please cleanup :)

File: packages/mobile/test/values.ts [hunk 9]
```diff
@@ -42,6 +42,7 @@ export const mockE164Number2 = '+12095559790'
 export const mockDisplayNumber2 = '(209) 555-9790'
 export const mockComment = 'Rent request for June, it is already late!!!'
 export const mockCountryCode = '+1'
+export const mockNestedObject = { '1': { '2': { '3': true, '4': false } } }
```
Comment (jmrossy) [inline packages/mobile/test/values.ts]: Remove this

**Review by jmrossy — APPROVED** [review]
Approved with one small comment, please fix before merging. 
Once you merge this, please copy over the updated db-rules.json to the firebase dbs and ping me when that's done :)

File: packages/mobile/src/firebase/firebase.ts [hunk 10]
```diff
@@ -118,13 +119,46 @@ export const writePaymentRequest = (paymentInfo: PaymentRequest) => async () =>
...
     .database()
-    .ref(`versions/${versionFSPath}`)
+    .ref('versions')
     .once('value')).val()
+  if (!versionsInfo || !versionsInfo.minVersion) {
+    return { deprecated, version }
+  }
+  const minVersion: string = versionsInfo.minVersion
+  deprecated = isDeprecatedVersion(version, minVersion)
+  return { version, deprecated }
```
Comment (martinvol) [inline packages/mobile/src/firebase/firebase.ts]: maybe the keys could be called: `current_ version` and `deprecated_ version`?

File: packages/mobile/src/firebase/firebase.ts [hunk 11]
```diff
@@ -118,13 +119,46 @@ export const writePaymentRequest = (paymentInfo: PaymentRequest) => async () =>
...
+}
+
+/*
+Get the Version deprecation information.
+@param version: string The version to check for deprecation
+@return: object { version: <VERSION>, deprecated: <BOOLEAN> }
+Firebase DB Format: 
+  (New) Add minVersion child to versions category with a string of the mininum version as string
+*/
 export async function getVersionInfo(version: string) {
```
Comment (martinvol) [inline packages/mobile/src/firebase/firebase.ts]: Can you type the output of this?

File: packages/mobile/src/firebase/firebase.test.ts [hunk 12]
```diff
@@ -87,13 +87,13 @@ describe(initializeCloudMessaging, () => {
 
 describe('Firebase version check', () => {
   it('Correctly check if version is deprecated', () => {
-    let isDeprecated: boolean = isDeprecatedVersion('1.5.0', '1.4.0')
+    let isDeprecated: boolean = isVersionBelowMinimum('1.5.0', '1.4.0')
```
Comment (martinvol) [inline packages/mobile/src/firebase/firebase.test.ts]: maybe you don't need to re-set that variable here and just 
`expect(isVersionBelowMinimum('1.5.0', '1.4.0')).toBe(false)`

Comment (jmrossy) [reply packages/mobile/src/firebase/firebase.test.ts]: Fixed

**Pull request merged**